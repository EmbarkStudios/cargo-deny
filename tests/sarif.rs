use cargo_deny::{
    diag::{DiagnosticCode, Severity},
    sarif::SarifLog,
    sarif_collector::SarifCollector,
};

#[test]
fn test_sarif_structure() {
    let sarif = SarifLog::default();

    assert_eq!(
        sarif.schema,
        "https://json.schemastore.org/sarif-2.1.0.json"
    );
    assert_eq!(sarif.version, "2.1.0");
    assert!(sarif.runs.is_empty());
}

#[test]
fn test_sarif_collector() {
    let mut collector = SarifCollector::default();

    // Add a test diagnostic
    collector.add_diagnostic(
        DiagnosticCode::Advisory(cargo_deny::advisories::Code::Vulnerability),
        Severity::Error,
        "Test vulnerability found".to_string(),
        "test.rs".to_string(),
        42,
    );

    let sarif = collector.generate_sarif();

    // Verify structure
    assert_eq!(sarif.version, "2.1.0");
    assert_eq!(sarif.runs.len(), 1);

    let run = &sarif.runs[0];
    assert_eq!(run.tool.driver.name, "cargo-deny");
    assert_eq!(run.tool.driver.version, env!("CARGO_PKG_VERSION"));

    // Verify we have rules
    assert_eq!(run.tool.driver.rules.len(), 1);
    let rule = &run.tool.driver.rules[0];
    assert_eq!(rule.id, "Advisory(Vulnerability)");
    assert_eq!(rule.name, "Advisory(Vulnerability)");
    assert!(rule.short_description.text.contains("Security"));

    // Verify we have results
    assert_eq!(run.results.len(), 1);
    let result = &run.results[0];
    assert_eq!(result.rule_id, "Advisory(Vulnerability)");
    assert_eq!(result.message.text, "Test vulnerability found");
    assert_eq!(result.level, "error");

    // Verify location
    assert_eq!(result.locations.len(), 1);
    let location = &result.locations[0];
    assert_eq!(location.physical_location.artifact_location.uri, "test.rs");
    assert_eq!(location.physical_location.region.start_line, 42);
}

#[test]
fn test_sarif_multiple_diagnostics() {
    let mut collector = SarifCollector::default();

    // Add multiple diagnostics of different types
    collector.add_diagnostic(
        DiagnosticCode::Advisory(cargo_deny::advisories::Code::Vulnerability),
        Severity::Error,
        "Vulnerability 1".to_string(),
        "lib.rs".to_string(),
        10,
    );

    collector.add_diagnostic(
        DiagnosticCode::License(cargo_deny::licenses::Code::Unlicensed),
        Severity::Warning,
        "Missing license".to_string(),
        "main.rs".to_string(),
        20,
    );

    collector.add_diagnostic(
        DiagnosticCode::Bans(cargo_deny::bans::Code::Banned),
        Severity::Error,
        "Banned crate".to_string(),
        "Cargo.toml".to_string(),
        30,
    );

    let sarif = collector.generate_sarif();

    // Should have 3 different rules
    assert_eq!(sarif.runs[0].tool.driver.rules.len(), 3);

    // Should have 3 results
    assert_eq!(sarif.runs[0].results.len(), 3);

    // Verify each result has correct severity
    let results = &sarif.runs[0].results;
    assert!(
        results
            .iter()
            .any(|r| r.message.text == "Vulnerability 1" && r.level == "error")
    );
    assert!(
        results
            .iter()
            .any(|r| r.message.text == "Missing license" && r.level == "warning")
    );
    assert!(
        results
            .iter()
            .any(|r| r.message.text == "Banned crate" && r.level == "error")
    );
}

#[test]
fn test_sarif_json_serialization() {
    let mut collector = SarifCollector::default();

    collector.add_diagnostic(
        DiagnosticCode::Advisory(cargo_deny::advisories::Code::Notice),
        Severity::Note,
        "Notice".to_string(),
        "test.rs".to_string(),
        1,
    );

    let sarif = collector.generate_sarif();

    // Should be serializable to JSON
    let json = serde_json::to_string(&sarif).expect("Should serialize to JSON");
    assert!(json.contains("\"$schema\""));
    assert!(json.contains("\"version\":\"2.1.0\""));
    assert!(json.contains("cargo-deny"));

    // Should be deserializable back
    let parsed: SarifLog = serde_json::from_str(&json).expect("Should deserialize from JSON");
    assert_eq!(parsed.version, sarif.version);
    assert_eq!(parsed.runs.len(), sarif.runs.len());
}

#[test]
fn test_sarif_empty_diagnostics() {
    // Edge case: no diagnostics collected
    let collector = SarifCollector::default();
    let sarif = collector.generate_sarif();

    // Should still produce valid SARIF structure
    assert_eq!(
        sarif.schema,
        "https://json.schemastore.org/sarif-2.1.0.json"
    );
    assert_eq!(sarif.version, "2.1.0");
    assert_eq!(sarif.runs.len(), 1);

    let run = &sarif.runs[0];
    assert_eq!(run.tool.driver.name, "cargo-deny");
    assert_eq!(run.tool.driver.version, env!("CARGO_PKG_VERSION"));

    // Should have no rules and no results
    assert!(run.tool.driver.rules.is_empty());
    assert!(run.results.is_empty());

    // Should be serializable to valid JSON
    let json = serde_json::to_string(&sarif).expect("Should serialize to JSON");
    assert!(json.contains("\"$schema\""));
    assert!(json.contains("\"version\":\"2.1.0\""));
    assert!(json.contains("\"results\":[]"));
    assert!(json.contains("\"rules\":[]"));
}
