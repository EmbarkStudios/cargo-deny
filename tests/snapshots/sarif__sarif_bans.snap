---
source: tests/sarif.rs
expression: s
---
{
  "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
  "runs": [
    {
      "results": [
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://sarif::sarif_bans.toml"
                },
                "region": {
                  "byteLength": 6,
                  "byteOffset": 99,
                  "message": "feature denied here",
                  "snippet": "simple",
                  "startLine": 4
                }
              }
            }
          ],
          "message": {
            "text": "feature 'simple' for crate 'features-galore = 0.1.0' is explicitly denied"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:feature-banned",
            "cargo-deny/krate": "path+file://{CWD}/tests/test_data/features-galore#features-galore@0.1.0"
          },
          "ruleId": "b:feature-banned"
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://sarif::sarif_bans.toml"
                },
                "region": {
                  "byteLength": 7,
                  "byteOffset": 208,
                  "message": "allowed features",
                  "snippet": "['ssh']",
                  "startLine": 6
                }
              }
            }
          ],
          "message": {
            "text": "feature 'default' for crate 'features-galore = 0.1.0' was not explicitly allowed"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:feature-not-explicitly-allowed",
            "cargo-deny/krate": "path+file://{CWD}/tests/test_data/features-galore#features-galore@0.1.0"
          },
          "ruleId": "b:feature-not-explicitly-allowed"
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://sarif::sarif_bans.toml"
                },
                "region": {
                  "byteLength": 7,
                  "byteOffset": 208,
                  "message": "allowed features",
                  "snippet": "['ssh']",
                  "startLine": 6
                }
              }
            }
          ],
          "message": {
            "text": "feature 'json' for crate 'features-galore = 0.1.0' was not explicitly allowed"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:feature-not-explicitly-allowed",
            "cargo-deny/krate": "path+file://{CWD}/tests/test_data/features-galore#features-galore@0.1.0"
          },
          "ruleId": "b:feature-not-explicitly-allowed"
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://sarif::sarif_bans.toml"
                },
                "region": {
                  "byteLength": 7,
                  "byteOffset": 208,
                  "message": "allowed features",
                  "snippet": "['ssh']",
                  "startLine": 6
                }
              }
            }
          ],
          "message": {
            "text": "feature 'simple' for crate 'features-galore = 0.1.0' was not explicitly allowed"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:feature-not-explicitly-allowed",
            "cargo-deny/krate": "path+file://{CWD}/tests/test_data/features-galore#features-galore@0.1.0"
          },
          "ruleId": "b:feature-not-explicitly-allowed"
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://sarif::sarif_bans.toml"
                },
                "region": {
                  "byteLength": 7,
                  "byteOffset": 208,
                  "message": "allowed features",
                  "snippet": "['ssh']",
                  "startLine": 6
                }
              }
            }
          ],
          "message": {
            "text": "feature 'zlib' for crate 'features-galore = 0.1.0' was not explicitly allowed"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:feature-not-explicitly-allowed",
            "cargo-deny/krate": "path+file://{CWD}/tests/test_data/features-galore#features-galore@0.1.0"
          },
          "ruleId": "b:feature-not-explicitly-allowed"
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://sarif::sarif_bans.toml"
                },
                "region": {
                  "byteLength": 14,
                  "byteOffset": 148,
                  "message": "feature denied here",
                  "snippet": "zlib-ng-compat",
                  "startLine": 5
                }
              }
            }
          ],
          "message": {
            "text": "feature 'zlib-ng-compat' for crate 'libssh2-sys = 0.2.23' is explicitly denied"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:feature-banned",
            "cargo-deny/krate": "libssh2-sys@0.2.23"
          },
          "ruleId": "b:feature-banned"
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://sarif::sarif_bans.toml"
                },
                "region": {
                  "byteLength": 5,
                  "byteOffset": 37,
                  "message": "banned here",
                  "snippet": "vcpkg",
                  "startLine": 2
                }
              }
            }
          ],
          "message": {
            "text": "crate 'vcpkg = 0.2.15' is explicitly banned"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:banned",
            "cargo-deny/krate": "vcpkg@0.2.15"
          },
          "ruleId": "b:banned"
        },
        {
          "level": "error",
          "locations": [],
          "message": {
            "text": "found 2 duplicate entries for crate 'windows-sys'"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:duplicate",
            "cargo-deny/krate": "windows-sys@0.36.1",
            "cargo-deny/krate1": "windows-sys@0.42.0"
          },
          "ruleId": "b:duplicate"
        },
        {
          "level": "error",
          "locations": [],
          "message": {
            "text": "found 2 duplicate entries for crate 'windows_aarch64_msvc'"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:duplicate",
            "cargo-deny/krate": "windows_aarch64_msvc@0.36.1",
            "cargo-deny/krate1": "windows_aarch64_msvc@0.42.0"
          },
          "ruleId": "b:duplicate"
        },
        {
          "level": "error",
          "locations": [],
          "message": {
            "text": "found 2 duplicate entries for crate 'windows_i686_gnu'"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:duplicate",
            "cargo-deny/krate": "windows_i686_gnu@0.36.1",
            "cargo-deny/krate1": "windows_i686_gnu@0.42.0"
          },
          "ruleId": "b:duplicate"
        },
        {
          "level": "error",
          "locations": [],
          "message": {
            "text": "found 2 duplicate entries for crate 'windows_i686_msvc'"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:duplicate",
            "cargo-deny/krate": "windows_i686_msvc@0.36.1",
            "cargo-deny/krate1": "windows_i686_msvc@0.42.0"
          },
          "ruleId": "b:duplicate"
        },
        {
          "level": "error",
          "locations": [],
          "message": {
            "text": "found 2 duplicate entries for crate 'windows_x86_64_gnu'"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:duplicate",
            "cargo-deny/krate": "windows_x86_64_gnu@0.36.1",
            "cargo-deny/krate1": "windows_x86_64_gnu@0.42.0"
          },
          "ruleId": "b:duplicate"
        },
        {
          "level": "error",
          "locations": [],
          "message": {
            "text": "found 2 duplicate entries for crate 'windows_x86_64_msvc'"
          },
          "partialFingerprints": {
            "cargo-deny/id": "b:duplicate",
            "cargo-deny/krate": "windows_x86_64_msvc@0.36.1",
            "cargo-deny/krate1": "windows_x86_64_msvc@0.42.0"
          },
          "ruleId": "b:duplicate"
        }
      ],
      "tool": {
        "driver": {
          "name": "cargo-deny",
          "rules": [
            {
              "defaultConfiguration": {
                "level": "error"
              },
              "fullDescription": {
                "text": ""
              },
              "help": {
                "markdown": "[bans check](https://embarkstudios.github.io/cargo-deny/checks/bans/index.html)\n[banned diagnostic](https://embarkstudios.github.io/cargo-deny/checks/bans/diags.html#banned)",
                "text": "For more information, see cargo-deny documentation"
              },
              "id": "b:banned",
              "name": "b:banned",
              "properties": {
                "precision": "high",
                "problem.severity": "error",
                "tags": [
                  "dependencies",
                  "supply-chain"
                ]
              },
              "shortDescription": {
                "text": "Detected an explicitly banned crate"
              }
            },
            {
              "defaultConfiguration": {
                "level": "error"
              },
              "fullDescription": {
                "text": ""
              },
              "help": {
                "markdown": "[bans check](https://embarkstudios.github.io/cargo-deny/checks/bans/index.html)\n[duplicate diagnostic](https://embarkstudios.github.io/cargo-deny/checks/bans/diags.html#duplicate)",
                "text": "For more information, see cargo-deny documentation"
              },
              "id": "b:duplicate",
              "name": "b:duplicate",
              "properties": {
                "precision": "high",
                "problem.severity": "error",
                "tags": [
                  "dependencies",
                  "supply-chain"
                ]
              },
              "shortDescription": {
                "text": "Detected two or more versions of the same crate"
              }
            },
            {
              "defaultConfiguration": {
                "level": "error"
              },
              "fullDescription": {
                "text": ""
              },
              "help": {
                "markdown": "[bans check](https://embarkstudios.github.io/cargo-deny/checks/bans/index.html)\n[feature-not-explicitly-allowed diagnostic](https://embarkstudios.github.io/cargo-deny/checks/bans/diags.html#feature-not-explicitly-allowed)",
                "text": "For more information, see cargo-deny documentation"
              },
              "id": "b:feature-not-explicitly-allowed",
              "name": "b:feature-not-explicitly-allowed",
              "properties": {
                "precision": "high",
                "problem.severity": "error",
                "tags": [
                  "dependencies",
                  "supply-chain"
                ]
              },
              "shortDescription": {
                "text": "A feature for a crate was not explicitly allowed"
              }
            },
            {
              "defaultConfiguration": {
                "level": "error"
              },
              "fullDescription": {
                "text": ""
              },
              "help": {
                "markdown": "[bans check](https://embarkstudios.github.io/cargo-deny/checks/bans/index.html)\n[feature-banned diagnostic](https://embarkstudios.github.io/cargo-deny/checks/bans/diags.html#feature-banned)",
                "text": "For more information, see cargo-deny documentation"
              },
              "id": "b:feature-banned",
              "name": "b:feature-banned",
              "properties": {
                "precision": "high",
                "problem.severity": "error",
                "tags": [
                  "dependencies",
                  "supply-chain"
                ]
              },
              "shortDescription": {
                "text": "An explicitly banned feature for a crate was set"
              }
            }
          ],
          "semanticVersion": "9.9.9",
          "version": "9.9.9"
        }
      }
    }
  ],
  "version": "2.1.0"
}
